kolibri/plugins/management/assets/src/actions.js

```diff
@@ -69,110 +81,145 @@ function _managePageTitle(title) {
-            // FacilityUserModel.save() will not send request to server.
-            responses.roles = [newRole];
-            store.dispatch('UPDATE_USERS', [responses]);
-          })
-          .catch(error => { coreActions.handleApiError(store, error); });
-        });
-      })
-      .catch(error => { coreActions.handleApiError(store, error); });
-    } else {
-    // role is learner and oldRole is admin or coach.
-      const OldRoleModel = RoleResource.getModel(oldRoldID);
-      OldRoleModel.delete().then(() => {
-        FacilityUserModel.save(payload).then(responses => {
-          // force role change because if the role is the only changing attribute
-          // FacilityUserModel.save() will not send request to server.
-          responses.roles = [];
-          store.dispatch('UPDATE_USERS', [responses]);
-        })
-        .catch(error => { coreActions.handleApiError(store, error); });
+function updateUser(store, stateUser) {
+  //payload needs username, fullname, and facility
+  const userID = stateUser.id;
+  const savedUserModel = FacilityUserResource.getModel(userID);
+  let savedUser = savedUserModel.attributes;
+  let roleAssigned = Promise.resolve(savedUserModel.attributes);
+  let changedValues = {};
+
+  // explicit checks for the only values that can be changed
+  if(stateUser.full_name !== savedUser.full_name)
+    changedValues.full_name = stateUser.full_name;
+
+  if(stateUser.username !== savedUser.username)
+    changedValues.username = stateUser.username;
+
+  if(stateUser.password !== savedUser.password)
+    changedValues.password = stateUser.password;
+
+  if(stateUser.facility !== savedUser.facility)
+    changedValues.facility = stateUser.facility;
+  
+  if(stateUser.kind !== _stateUser(savedUser).kind){
+    // assumes there's no previous roles to delete at first
+    let handlePreviousRoles = Promise.resolve();
+  
+    if(savedUser.roles.length){
+      let roleDeletes = [];
+      savedUser.roles.forEach(role => {
+        roleDeletes.push(RoleResource.getModel(role.id).delete());
       });
+
+      
+      // delete the old role models if this was not a learner
+      handlePreviousRoles = Promise.all(roleDeletes).then(responses => {
+
+        // to avoid having to make an API call, clear manually (used in assign)
+        savedUser.roles = [];
+
+        // gives access to delete responses if wanted
+        return responses;
+      
+      // models could not be deleted
+      }, error => error);
     }
-  } else {
-  // the role is not changed
-    FacilityUserModel.save(payload).then(responses => {
-      store.dispatch('UPDATE_USERS', [responses]);
-    })
-    .catch(error => { coreActions.handleApiError(store, error); });
+
+    // then assign the new role
+    roleAssigned = new Promise((resolve,reject) => {
+
+      // Take care of previous roles if necessary (will autoresolve if not)
+      handlePreviousRoles.catch(error => reject(error));
```

Comment:
This comment string was here when I first wrote the PR. I just never confirmed my "review" to get all my comments in. I was wondering what "pending" meant. :|